### PR TITLE
Add pytorch tensor's cummax method

### LINKED ISSUE #22325: cummax

### PR DESCRIPTION
**Pull request by heisguyy**
This commit adds the cummax method implemented in the miscellaneous_ops as a method in the pytorch frontend definition of tensors. This allows ivy users to call the cummax method of a pytorch tensor using the ivy's namespace instead of pytorch's namespace.

Closes #22325

Socials: heisguyy

### CODE REVIEW

**Comment by heisguyy** [comment]
**Note to reviewer**
After initially writing the method and the test, I was faced with this error.

"""
/ivy_tests/test_ivy/test_frontends/test_torch/test_tensor-py::test_torch_tensor_cummax
[cpu-numpy-False-Falsel Failed: [undefined]AssertionError: returned dtype = int64,
ground-truth returned dtype = int8
"""

My thoughts: Ivy created the arrays with int64 dtype and pytorch with int8 hence the different.
My proposed solution: Add dtype parameter to ensure ivy and pytorch creates the same type of arrays.

After I did this, I get this error
"""
/ivy_tests/test_ivy/test_frontends/test_torch/test
_tensor-py::test_torch_tensor_cummax
[cpu-numpy-False-False] Failed: [undefined]TypeError: cummax) received an invalid combination of arguments -
got (dtype=torch.dtype, dim=int, ), but expected one of:
* (int dim)
* (name dim)
"""

I am currently stuck and any advise on how to proceed would be helpful. Thank you for your time.

File: ivy/functional/frontends/torch/miscellaneous_ops.py [hunk 1]
```diff
@@ -122,8 +122,10 @@ def cross(input, other, dim=None, *, out=None):
 
 
 @to_ivy_arrays_and_back
-def cummax(input, dim, *, out=None):
-    return ivy.cummax(input, axis=dim, out=out)
+def cummax(input, dim, *, dtype=None, out=None):
+    if not dtype and "int" in input.dtype:
+        dtype = ivy.int64
+    return ivy.cummax(input, axis=dim, dtype=dtype, out=out)
```
Comment (jshepherd01) [inline ivy/functional/frontends/torch/miscellaneous_ops.py:128]: [`torch.cummax`](https://pytorch.org/docs/stable/generated/torch.cummax.html) doesn't take a dtype argument, and whatever we do we need to stick to the original function signature

File: ivy/functional/frontends/torch/tensor.py [hunk 2]
```diff
@@ -1846,6 +1846,12 @@ def index_fill(self, dim, index, value):
         arr = torch_frontend.moveaxis(self, 0, dim)
         return arr
 
+    @with_unsupported_dtypes(
+        {"2.0.1 and below": ("uint8", "uint16", "bfloat16", "float16")}, "torch"
+    )
+    def cummax(self, dim, dtype=None):
+        return torch_frontend.cummax(self, dim, dtype=dtype)
```
Comment (jshepherd01) [inline ivy/functional/frontends/torch/tensor.py:1853]: `torch.Tensor.cummax` also doesn't take a `dtype` argument

File: ivy/functional/frontends/torch/tensor.py [hunk 3]
```diff
@@ -1846,6 +1846,12 @@ def index_fill(self, dim, index, value):
         arr = torch_frontend.moveaxis(self, 0, dim)
         return arr
 
+    @with_unsupported_dtypes(
+        {"2.0.1 and below": ("uint8", "uint16", "bfloat16", "float16")}, "torch"
+    )
```
Comment (jshepherd01) [inline ivy/functional/frontends/torch/tensor.py:1851]: How did you arrive at this list of unsupported types? The function in `miscellaneous_ops` seems to support these, so the instance method ideally should too

File: ivy_tests/test_ivy/test_frontends/test_torch/test_tensor.py [hunk 4]
```diff
@@ -12437,3 +12437,47 @@ def test_torch_triu_(
...
+    dtype_value=helpers.dtype_and_values(
+        available_dtypes=helpers.get_dtypes("numeric"),
+        shape=st.shared(helpers.get_shape(min_num_dims=1), key="shape"),
+    ),
+    dim=helpers.get_axis(
+        shape=st.shared(helpers.get_shape(), key="shape"),
+        allow_neg=True,
+        force_int=True,
+    ),
+    dtypes=_dtypes(),
```
Comment (jshepherd01) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_tensor.py:12456]: You can remove the `dtypes` argument from here too

**Review by jshepherd01 — CHANGES_REQUESTED** [review]
I've suggested some changes. As far as the bug you pointed out, it seems like that's also an issue with the original frontend implementation of `cummax`, basically the function returns two arrays and torch prefers to keep one of them the same type as the input whereas ivy upcasts it to match the other output. I think to resolve the issue, you'd need to modify the `cummax` function in `miscellaneous_ops.py` to convert the first output array to the same type as the input

File: ivy/functional/frontends/torch/tensor.py (hunk 3 above)
Comment (heisguyy) [reply ivy/functional/frontends/torch/tensor.py:1851]: I actually got most of this from the backend implementation in ivy/functional/backends/torch/experimental/statistical.py. The other came up as a solution to a error I was facing while testing.

Comment (heisguyy) [reply ivy/functional/frontends/torch/tensor.py:1851]: ok, thank you. I will do that rn.

**Comment by heisguyy** [comment]
I have made all the changes you recommended. The unsupported datatypes are the ones that cause the test function to fail.  All tests are currently working apart from the one with numpy backend which gives this error. 
`
/ivy_tests/test_ivy/test_frontends/test_torch/test_tensor-py::test_torch_tensor_cummax [cpu-numpy-False-Falsel Failed:
[undefined]AssertionError: returned shape = (1,), ground-truth returned shape = (1, 1)
`
I have tried to creating a tensor with pytorch and ivy(numpy backend) and calling the cummax method on it, but I am unable to see the different in shape as the test error denotes.

**Review by jshepherd01 — COMMENTED** [review]
There's a couple of extra tests that you seem to have added that I'm not sure why.

About the bug you've seen, I think that's a problem with the underlying implementation of `cummax` which we (Ivy) probably need to fix, so I'm happy to overlook that for the PR

File: ivy_tests/test_ivy/test_frontends/test_torch/test_tensor.py [hunk 5]
```diff
@@ -554,6 +554,40 @@ def test_torch___array_wrap__(
...
+        local_importer = ivy_backend.utils.dynamic_import
+        function_module = local_importer.import_module("ivy.functional.frontends.torch")
+        ret = function_module.tensor(x[0]).__array_wrap__(x[1])
+        assert isinstance(ret, function_module.Tensor)
+    helpers.value_test(
+        ret_np_flat=np.array(ret.ivy_array).ravel(),
+        ret_np_from_gt_flat=ret_gt.numpy().ravel(),
+        ground_truth_backend="torch",
+        backend=backend_fw,
+    )
```
[the remark this answers is not in the document]
Comment (heisguyy) [reply ivy_tests/test_ivy/test_frontends/test_torch/test_tensor.py:588]: Oh, that was from one main repository. When I pulled, I had to resolve a merge conflict that's why it in one of my commit

File: ivy_tests/test_ivy/test_frontends/test_torch/test_tensor.py [hunk 6]
```diff
@@ -1447,6 +1481,37 @@ def test_torch__array__(
...
+        local_importer = ivy_backend.utils.dynamic_import
+        function_module = local_importer.import_module("ivy.functional.frontends.torch")
+        ret = function_module.tensor(x[0]).__array__(dtype[0])
+
+    helpers.value_test(
+        ret_np_flat=ret.ravel(),
+        ret_np_from_gt_flat=ret_gt.ravel(),
+        ground_truth_backend="torch",
+        backend=backend_fw,
+    )
```
[the remark this answers is not in the document]
Comment (heisguyy) [reply ivy_tests/test_ivy/test_frontends/test_torch/test_tensor.py:1512]: This is also from resolving the merge conflicts. It's originally from someone else in the main repo.

File: ivy_tests/test_ivy/test_frontends/test_torch/test_tensor.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jshepherd01) [reply ivy_tests/test_ivy/test_frontends/test_torch/test_tensor.py:588]: Ah, I see. It seems like you've accidentally duplicated it though (another copy of the same test function is directly above it in the same file)

File: ivy_tests/test_ivy/test_frontends/test_torch/test_tensor.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jshepherd01) [reply ivy_tests/test_ivy/test_frontends/test_torch/test_tensor.py:1512]: Same with this one, it's a duplicate of the function directly above it

File: ivy_tests/test_ivy/test_frontends/test_torch/test_tensor.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (heisguyy) [reply ivy_tests/test_ivy/test_frontends/test_torch/test_tensor.py:588]: Oh, my bad. I will do that right now.

File: ivy_tests/test_ivy/test_frontends/test_torch/test_tensor.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (heisguyy) [reply ivy_tests/test_ivy/test_frontends/test_torch/test_tensor.py:1512]: On it.

**Review by jshepherd01 — CHANGES_REQUESTED** [review]
When I run the tests locally, both float16 and uint8 seem to be accepted dtypes. Could you please update those, and then I'd be happy to merge it

**Pull request merged**